File: problems/36/36.valid-sudoku.go
/*
 * @lc app=leetcode id=36 lang=golang
 *
 * [36] Valid Sudoku
 */

package solution

import "strconv"

// @lc code=start
func isValidSudoku(board [][]byte) bool {
	// check rows
	for i := 0; i < 9; i++ {
		curr := make([]bool, 9)
		for j := 0; j < 9; j++ {
			if string(board[i][j]) == "." {
				continue
			}
			num, _ := strconv.Atoi(string(board[i][j]))
			if !curr[num-1] {
				curr[num-1] = true
			} else {
				return false
			}
		}
	}

	// check cols
	for j := 0; j < 9; j++ {
		curr := make([]bool, 9)
		for i := 0; i < 9; i++ {
			if string(board[i][j]) == "." {
				continue
			}
			num, _ := strconv.Atoi(string(board[i][j]))
			if !curr[num-1] {
				curr[num-1] = true
			} else {
				return false
			}
		}
	}

	// check 3x3
	for i := 0; i < 9; i += 3 {
		for j := 0; j < 9; j += 3 {
			curr := make([]bool, 9)
			for ii := i; ii < i+3; ii++ {
				for jj := j; jj < j+3; jj++ {
					if string(board[ii][jj]) == "." {
						continue
					}
					num, _ := strconv.Atoi(string(board[ii][jj]))
					if !curr[num-1] {
						curr[num-1] = true
					} else {
						return false
					}
				}
			}
		}
	}

	return true
}

// @lc code=end

File: problems/36/36.valid-sudoku.py
#
# @lc app=leetcode id=36 lang=python3
#
# [36] Valid Sudoku
#


# @lc code=start
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # valid if
        # 1. each row contains digits 1-9 without repetition
        # 2. each col contains digits 1-9 without repetition
        # 3. each 3x3 contains digits 1-9 without repetition

        ROWS, COLS = 9, 9

        # check rows
        for r in range(ROWS):
            seen = set()
            for c in range(COLS):
                curr = board[r][c]
                if curr == ".":
                    continue
                if curr in seen:
                    return False
                seen.add(curr)

        # check cols
        for c in range(COLS):
            seen = set()
            for r in range(ROWS):
                curr = board[r][c]
                if curr == ".":
                    continue
                if curr in seen:
                    return False
                seen.add(curr)

        # check 3x3
        for c in range(0, COLS, 3):
            for r in range(0, ROWS, 3):
                seen = set()
                for i in range(9):
                    curr = board[r + (i // 3)][c + (i % 3)]
                    if curr == ".":
                        continue
                    if curr in seen:
                        return False
                    seen.add(curr)

        return True


# @lc code=end
